Feature/api integration

mohammedyashour opened the pull request:
This PR includes the following enhancements:

     Implemented LocationApiClient to fetch city coordinates and current location using IP.

     Updated WeatherApiClient to support fetching weather by IP (getWeatherByIp).

     Created GetWeatherByIpUseCase to abstract weather retrieval via IP.

     Updated WeatherDataSource and WeatherRepository interfaces to support the new use case.

     Improved error handling for invalid city or country using custom exception LocationNotFoundException.

     Enhanced main function to demonstrate both use cases: by IP and by city/country.

     Ensured graceful fallback with try/catch and proper messages for user feedback

mohammedyashour committed:
create feature/api-integration branch

mohammedyashour committed:
Merge branch 'develop' into feature/api-integration

# Conflicts:
#	build.gradle.kts

mohammedyashour committed:
feat: implement weather API integration and use case

mohammedyashour committed:
implement weather feature tests and apply general improvements

elsabagh committed:
Add a feature to retrieve the weather by current location or entering the city name and create a test case

elsabagh committed:
Merge remote-tracking branch 'origin/feature/api-integration' into feature/api-integration

# Conflicts:
#	build.gradle.kts
#	src/main/kotlin/data/weather/datasource/WeatherApiClient.kt
#	src/main/kotlin/domain/exception/Exceptions.kt
#	src/test/kotlin/data/weather/datasource/WeatherApiClientTest.kt
#	src/test/kotlin/domain/usecase/GetWeatherUseCaseTest.kt

elsabagh committed:
update branch and solve some conflict

elsabagh committed:
edit exception and location api

elsabagh committed:
refactor location api and edit test case

AbdoMagdey24 commented on src/main/kotlin/data/location/dataSource/LocationApiClient.kt:
```diff
@@ -0,0 +1,23 @@
+) : LocationDataSource {
+
+    override suspend fun getCityCoordinates(city: String, country: String): LocationDto? {
+        val geoUrl = "https://geocoding-api.open-meteo.com/v1/search?name=$city,$country&count=1"
```
we can do a companion object for base_url and do something like this 
client.get {
            url(BASE_URL)
            parameter("name", "$cityName,$country")
        }

AbdoMagdey24 commented on src/main/kotlin/data/remote/HttpClientProvider.kt:
```diff
@@ -0,0 +1,23 @@
+import io.ktor.serialization.kotlinx.json.*
+import kotlinx.serialization.json.Json
+
+object HttpClientProvider {
```
Where is the usage of this provider ?

AbdoMagdey24 commented on src/main/kotlin/data/weather/datasource/WeatherApiClient.kt:
```diff
@@ -1,11 +1,39 @@
+
+
+    private suspend fun getWeatherByLocation(location: LocationDto): WeatherDto {
+        val weatherUrl = "https://api.open-meteo.com/v1/forecast" +
```
also here we can use companion object

AbdoMagdey24 commented on src/main/kotlin/data/weather/mapper/WeatherMapper.kt:
```diff
@@ -7,22 +7,58 @@ import org.damascus.domain.model.WeatherUnit
     )
+
+}
+fun getWeatherDescription(code: Int): String {
```
why there is no usage for this function unless in testing ?
also you will need to cover all cases to pass from Jacoco branch coverage

AbdoMagdey24 commented on src/main/kotlin/domain/exception/Exceptions.kt:
```diff
@@ -1,3 +1,5 @@
 package org.damascus.domain.exception
 
-class WeatherNotFoundException : Exception("No weather found")
```
Is WeatherNotFoundException useless so that you removed it or what ?

AbdoMagdey24 commented on src/main/kotlin/domain/exception/Exceptions.kt:
```diff
@@ -1,3 +1,5 @@
\ No newline at end of file
+import org.koin.core.logger.MESSAGE
+
+class LocationNotFoundException(message: MESSAGE) : Exception(message)
```
why we need MESSAGE logger ? why not string

mohammedyashour committed:
delete HttpClientProvider.kt as it's no longer needed
enhance LocationApiClient,WeatherApiClient  reliability and maintainability
increase WeatherMapperTest.kt Coverage

mohammedyashour committed:
improve WeatherMapper test cases to achieve 100% branch coverage

yusufnasserdev commented on src/main/kotlin/data/location/dataSource/LocationApiClient.kt:
```diff
@@ -0,0 +1,31 @@
+    private val client: HttpClient
+) : LocationDataSource {
+
+    companion object {
```
It's better to make the companion object private if you want to prevent it from being used outside the class, instead of privatizing all its members.

elsabagh replied:
done

yusufnasserdev commented on src/main/kotlin/domain/usecase/GetWeatherUseCase.kt:
```diff
@@ -1,9 +1,14 @@
+
+
 class GetWeatherUseCase(
     private val repository: WeatherRepository
```
Give the `WeatherRepository` a better name to describe what repository this is.

elsabagh replied:
done

yusufnasserdev commented on src/main/kotlin/data/weather/dto/CurrentWeatherUnits.kt:
```diff
@@ -2,14 +2,13 @@ package org.damascus.data.weather.dto
-    val weatherCode: String
-)
\ No newline at end of file
+    @SerialName("weathercode") val weatherCode: String
```
Please avoid having the annotation in the same line as the member variable.

elsabagh replied:
done

yusufnasserdev commented on src/main/kotlin/data/weather/dto/CurrentWeather.kt:
```diff
@@ -2,14 +2,13 @@ package org.damascus.data.weather.dto
-    val weatherCode: Int
-)
\ No newline at end of file
+    @SerialName("weathercode") val weatherCode: Int
```
Please avoid having the annotation in the same line as the member variable.

elsabagh replied:
done

yusufnasserdev commented on src/main/kotlin/data/location/dto/AutoLocationDto.kt:
Please rename this, `Auto` doesn't provide meaning in this context.

elsabagh replied:
done

yusufnasserdev commented on src/main/kotlin/data/weather/datasource/WeatherApiClient.kt:
```diff
@@ -1,11 +1,42 @@
+            parameter("current_weather", true)
+            parameter("timezone", "auto")
+        }.body()
+    }}
```
Refactor code, please utilize intellij `reformat code` `ctrl + alt + L`

elsabagh replied:
done

yusufnasserdev commented on src/main/kotlin/data/weather/dto/CurrentWeatherUnits.kt:
```diff
@@ -2,14 +2,13 @@ package org.damascus.data.weather.dto
 import kotlinx.serialization.Serializable
-
 @Serializable
 data class CurrentWeatherUnits(
```
Please use better naming, this should follow the same naming as the rest of the package `dto`

yusufnasserdev commented on src/main/kotlin/data/weather/dto/CurrentWeather.kt:
```diff
@@ -2,14 +2,13 @@ package org.damascus.data.weather.dto
 import kotlinx.serialization.Serializable
-
 @Serializable
 data class CurrentWeather(
```
Please use better naming, this should follow the same naming as the rest of the package `dto`

yusufnasserdev commented on src/main/kotlin/data/weather/mapper/WeatherMapper.kt:
```diff
@@ -7,22 +7,58 @@ import org.damascus.domain.model.WeatherUnit
         ),
         units = WeatherUnit(
             temperatureUnit = currentWeatherUnits.temperature.ifBlank { "°C" },
-            windSpeedUnit = currentWeatherUnits.windSpeed.ifBlank { "km/h" },
-            windDirectionUnit = currentWeatherUnits.windDirection.ifBlank { "°" }
```
Why removed? please reply with an explanation

mohammedyashour replied:
It was removed by mistake while testing their functionality in the code
 I forgot to return the rest
will fix it

yusufnasserdev commented on src/test/kotlin/data/location/dataSource/LocationApiClientTest.kt:
```diff
@@ -0,0 +1,84 @@
+
+class LocationApiClientTest {
+
+    private val json = Json { ignoreUnknownKeys = true }
```
This should be in a `@BeforeEach`

elsabagh replied:
done

yusufnasserdev commented on src/test/kotlin/data/location/dataSource/LocationApiClientTest.kt:
```diff
@@ -0,0 +1,84 @@
+        val result = api.getCityCoordinates("Cairo", "EG")
+
+        assertNotNull(result)
+        assertEquals(30.0, result.latitude)
+        assertEquals(31.0, result.longitude)
```
This should be one assert.

elsabagh replied:
done

yusufnasserdev commented on src/test/kotlin/data/weather/mapper/WeatherMapperTest.kt:
```diff
@@ -0,0 +1,101 @@
+            99 to "Thunderstorm with heavy hail"
+        )
+
+        for ((code, expected) in cases) {
+            val result = getWeatherDescription(code)
+            assertThat(result).isEqualTo(expected)
+        }
+    }
```
Using `if` and `for` in test code is frowned upon, you shouldn't introduce logic within the test code.

Instead, you should use a parameterized test.

elsabagh replied:
done

yusufnasserdev commented on src/test/kotlin/domain/usecase/GetWeatherByIpUseCaseTest.kt:
```diff
@@ -0,0 +1,84 @@
+    @BeforeEach
+    fun setup() {
+        repository = mockk()
+        useCase = GetWeatherByIpUseCase(repository)
```
Give this a better name please.

elsabagh replied:
done

yusufnasserdev commented on src/test/kotlin/domain/usecase/GetWeatherByIpUseCaseTest.kt:
```diff
@@ -0,0 +1,84 @@
+        useCase = GetWeatherByIpUseCase(repository)
+    }
+
+    @Test
+    fun `should return weather info from IP location`() = runTest {
+        // Given
+        val expected = dummyWeatherInfo()
+        coEvery { repository.getWeatherByIp() } returns expected
+
+        // When
+        val result = useCase()
+
+        // Then
+        assertEquals(expected, result)
+    }
+
+    @Test
+    fun `should return correct temperature from IP location`() = runTest {
+        // Given
+        val expected = dummyWeatherInfo()
+        coEvery { repository.getWeatherByIp() } returns expected
+
+        // When
+        val result = useCase()
+
+        // Then
+        assertEquals(26.0, result.weather.temperature)
+    }
```
Parameterized test.

elsabagh replied:
I removed the temperature test because it is repetitive as I am considered a user in the first case

yusufnasserdev commented on src/test/kotlin/data/weather/datasource/WeatherApiClientTest.kt:
```diff
@@ -0,0 +1,127 @@
+    }
+
+    @Test
+    fun `should return throw when city is invalid`() = runTest {
```
`should throw`, `throw` isn't a return type.

elsabagh replied:
done

yusufnasserdev commented on src/test/kotlin/data/weather/datasource/WeatherApiClientTest.kt:
```diff
@@ -0,0 +1,127 @@
+
+    @Test
+    fun `should return Weather data when city is valid`() = runTest {
+        locationDataSource = object : LocationDataSource {
+            override suspend fun getCityCoordinates(city: String, country: String) = LocationDto(30.0, 31.0)
+            override suspend fun getCurrentLocation() = null
+        }
```
This should be done using `mockk`

elsabagh replied:
done

elsabagh committed:
modify the name of some files, format the code, and modify the test case.

yusufnasserdev commented on src/test/kotlin/data/weather/mapper/WeatherMapperTest.kt:
```diff
@@ -0,0 +1,151 @@
+
+        val info = dto.toWeatherInfo()
+
+        assertThat(info.timezone).isEqualTo("Asia/Gaza")
+        assertThat(info.weather.isDay).isTrue()
+        assertThat(info.units.temperatureUnit).isEqualTo("°C")
```
Please avoid having multiple assertions per test.

mohammedyashour replied:
done

yusufnasserdev commented on src/test/kotlin/data/weather/mapper/WeatherMapperTest.kt:
Please follow `Given/When/Then` comments pattern.

mohammedyashour replied:
done

yusufnasserdev commented on src/test/kotlin/domain/usecase/GetWeatherByIpUseCaseTest.kt:
```diff
@@ -0,0 +1,72 @@
+        assertEquals(expected, result)
+    }
+
+    @Test
+    fun `should return throws exception when fail`() = runTest {
+        // Given 
+        coEvery { repository.getWeatherByIp() } throws LocationNotFoundException("Could not determine location from IP")
+
+        // When & Then
+        assertFailsWith<LocationNotFoundException> {
+        weatherByIpUseCase()
+        }
+    }
```
It's either return or throws.

mohammedyashour replied:
done

mohammedyashour committed:
update GetWeatherByIpUseCaseTest
WeatherMapperTest.kt

elsabagh committed:
add a test case serialize data and test the real api

elsabagh committed:
create a test case to clothes dataSource

elsabagh commented:
please check change @yusufnasserdev  @Asmaa7071

Al-Taie commented on src/main/kotlin/data/location/dataSource/LocationApiClient.kt:
```diff
@@ -0,0 +1,31 @@
+    private val client: HttpClient
+) : LocationDataSource {
+
+    private companion object {
+        private const val GEO_BASE_URL = "https://geocoding-api.open-meteo.com/v1/search"
+        private const val IP_BASE_URL = "http://ip-api.com/json/"
+    }
```
move private things to bottom

elsabagh replied:
done

Al-Taie commented on src/main/kotlin/data/location/dto/IpLocationDto.kt:
```diff
@@ -0,0 +1,12 @@
+
+@Serializable
+data class IpLocationDto(
+    @SerialName("lat")
+    val lat: Double,
+    @SerialName("lon")
+    val lon: Double
```
You can change the name in the data class while you are already mentioned the api names in the anotiation

elsabagh replied:
done

Al-Taie commented on src/main/kotlin/data/weather/datasource/WeatherApiClient.kt:
```diff
@@ -1,11 +1,44 @@
+    private val client: HttpClient,
+    private val locationDataSource: LocationDataSource
+) : WeatherDataSource {
+    companion object {
+        private const val WEATHER_BASE_URL = "https://api.open-meteo.com/v1/forecast"
+    }
```
move private things to bottom
```suggestion
    private companion object {
       const val WEATHER_BASE_URL = "https://api.open-meteo.com/v1/forecast"
    }
```

elsabagh replied:
done

Al-Taie commented on src/main/kotlin/data/weather/datasource/WeatherApiClient.kt:
```diff
@@ -1,11 +1,44 @@
+        val ipLocation = locationDataSource.getCurrentLocation()
+            ?: throw LocationNotFoundException("Could not determine location from IP")
+
+        return getWeatherByLocation(LocationDto(ipLocation.lat, ipLocation.lon))
```
you can make this as an extension function.

elsabagh replied:
done

Al-Taie commented on src/main/kotlin/data/weather/mapper/WeatherMapper.kt:
```diff
@@ -7,22 +7,58 @@ import org.damascus.domain.model.WeatherUnit
 
 fun WeatherDto.toWeatherInfo(): WeatherInfo {
```
you can name it `toDomain()`

elsabagh replied:
done

Al-Taie commented on src/test/kotlin/data/location/dataSource/LocationApiClientTest.kt:
We should not test the api, we should test our logic only.

Al-Taie commented on src/test/kotlin/data/weather/datasource/WeatherApiClientTest.kt:
[Check this comment](https://github.com/damascus-squad/Fit-Sync/pull/11/files#r2081864111)

elsabagh committed:
refactor name files, add SerializeName to the rest of the elements, reformat code and add extension fun toLocationDto

elsabagh committed:
add SerializeName to the rest of the elements

elsabagh committed:
refactor mapper back ifBlank
